File: OJ/leetcode/python/79_word_search.py
from typing import List
from typing import Tuple

class Solution:
    def findNeighbors(self, i: int, j: int, valid: List[List[bool]]) -> List[Tuple]:
        neighbors = []
        m, n = len(valid), len(valid[0])
        if i - 1 >= 0 and valid[i - 1][j]:
            neighbors.append((i - 1, j))
        if j - 1 >= 0 and valid[i][j - 1]:
            neighbors.append((i, j - 1))
        if i + 1 < m and valid[i + 1][j]:
            neighbors.append((i + 1, j))
        if j + 1 < n and valid[i][j + 1]:
            neighbors.append((i, j + 1))
        return neighbors

    def dfs(self, row: int, col: int, word: str, \
            valid: List[bool], board: List[List[str]]) -> bool:
        if len(word) == 0:
            return True
        neighbors = self.findNeighbors(row, col, valid)
        for neighbor in neighbors:
            i, j = neighbor
            if word[0] == board[i][j]:
                valid[i][j] = False
                if self.dfs(i, j, word[1:], valid, board):
                    return True
                valid[i][j] = True
        return False

    def exist(self, board: List[List[str]], word: str) -> bool:
        valid = [[True for _ in board[0]] for _ in board]
        m, n = len(board), len(board[0])
        for i in range(m):
            for j in range(n):
                if board[i][j] == word[0]:
                    valid[i][j] = False
                    ans = self.dfs(i, j, word[1:], valid, board)
                    if ans:
                        return True
                    valid[i][j] = True
        return False
                     

solution = Solution()

board = [ \
    ["C","A","A"], \
    ["A","A","A"], \
    ["B","C","D"]  \
]
word = "AAB"
print(solution.exist(board, word))

board = [ \
    ["A","B","C","E"], \
    ["S","F","C","S"], \
    ["A","D","E","E"]  \
]
word = "ABCCED"
print(solution.exist(board, word))

board = [ \
    ["A","B","C","E"], \
    ["S","F","C","S"], \
    ["A","D","E","E"]  \
]
word = "SEE"
print(solution.exist(board, word))

board = [ \
    ["A","B","C","E"], \
    ["S","F","C","S"], \
    ["A","D","E","E"]  \
]
word = "ABCB"
print(solution.exist(board, word))

board = [["a","a"]]
word = "aaa"
print(solution.exist(board, word))

File: competitive_programming/leetcode/cpp/79_word_search.cpp
#include <iostream>
#include <vector>
#include <utility>
#include <string>
using namespace std;

vector<pair<int, int>> getNeighbors(int i, int j, const vector<vector<char>>& board) {
    int m = board.size();
    int n = board[0].size();
    vector<pair<int, int>> neighbors;
    if (i - 1 >= 0) {
        neighbors.push_back(make_pair(i - 1, j));
    }
    if (i + 1 < m) {
        neighbors.push_back(make_pair(i + 1, j));
    }
    if (j - 1 >= 0) {
        neighbors.push_back(make_pair(i, j - 1));
    }
    if (j + 1 < n) {
        neighbors.push_back(make_pair(i, j + 1));
    }
    return neighbors;
}

bool dfs(int row, int col, int wi, vector<vector<bool>>& valid, 
         const vector<vector<char>>& board, string& word) {
    if (wi == word.size()) {
        return true;
    }
    vector<pair<int, int>> neighbors = getNeighbors(row, col, board);
    for (pair<int, int> neighbor: neighbors) {
        int i = neighbor.first;
        int j = neighbor.second;
        if (valid[i][j] and word[wi] == board[i][j]) {
            valid[i][j] = false;
            bool ans = dfs(i, j, wi + 1, valid, board, word);
            valid[i][j] = true;
            if (ans) {
                return true;
            }
        }
    }
    return false;
}

bool exist(vector<vector<char>>& board, string word) {
    int m = board.size();
    int n = board[0].size();
    vector<vector<bool>> valid(m, vector<bool>(n, true));
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            if (board[i][j] == word[0]) {
                valid[i][j] = false;
                bool ans = dfs(i, j, 1, valid, board, word);
                valid[i][j] = true;
                if (ans) {
                    return true;
                }
            }
        }
    }
    return false;
}

int main() {
    vector<vector<char>> board = {
        {'A','B','C','E'},
        {'S','F','C','S'},
        {'A','D','E','E'}
    };
    string word = "ABCCED";
    cout << exist(board, word) << "\n";
}